### ISSUE TITLE
xFLAGS in MPICH2, MVAPICH2 and potentially other MPICH2 derivatives 

### ISSUE THREAD

**Issue by damianam** [hpcugent/easybuild-easyconfigs#2334]
MPICH2 and its derivatives have an obscure "feature" that can potentially affect all the sites that use EB. As can be checked in the README file, popular environment variables like CFLAGS (among others) are used to compile the runtime, but they are added in the configure step to mpixxx wrappers (mpicc, mpicxx, etc). That means that compiling the MPI runtime with -O2 -g results in users compiling their applications with potentially conflicting flags (e.g.: mpicc -O2 -g -O3, with -O3 being the desired flag, and -O2 -g being transparently and inadvertently added by the MPI wrapper). Other flags can have more obscure consequences.

This is true for all (or at least most) versions of MPICH2. Check section 3 of these files:

http://www.mpich.org/static/downloads/1.5/mpich2-1.5-README.txt
http://www.mpich.org/static/downloads/3.2/mpich-3.2-README.txt

IMO this should be fixed due to its potential problems.

MVAPICH2 has also the same issue, inhered from MPICH2.

As a side note, MVAPICH2 is distributed sometimes as an RPM binary. It seems like at least some of these RPMs (like MVAPICH2-GDR 2.1 and 2.2a) are affected similarly, probably because the packager had xFLAGS defined at compile time. The list of flags passed to applications compiled with MPI in this case is:

-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic

whose impact depends on the backend compiler behaviour (whether -O2 or -O3 has precedences, conflicting -mtune settings, etc).

**Comment by boegel** [comment]
Waw, this is... interesting.

Here's what I have:

```
$ module load MPICH/3.0.4-GCC-4.8.1
$ grep '\-O' $EBROOTMPICH/bin/mpicc
CFLAGS="-O2 -march=native "
```

and

```
$ module load MVAPICH2/1.9-iccifort-2011.13.367
$ grep '\-O' $EBROOTMVAPICH2/bin/mpicc
CFLAGS="-fPIC -O2 -xHost -ftz -fp-speculation=safe -fp-model source "
```

I think in general, the optimisation flags you specify will 'win' over the ones embedded in the MPI compiler wrappers, but still, this is a mess.

So, basically, we need to use `$MPICH2LIB_CFLAGS` rather than `$CFLAGS` when building MPICH v2.x and v3.x, and whatever the equivalent is for MVAPICH2.

This is a very good reason to add an easyblock for MPICH, and enhance the existing MVAPICH2 easyblock, both to make sure that `$CFLAGS` is undefined and `$MPICH2LIB_CFLAGS` is used instead, *and* to check that `mpicc` doesn't contain hardcoded values.

It does not apply to Intel MPI however (which is MPICH (1.x) based iirc), as far as I can tell.

The surprising thing to me is that `mpicc` is totally ignoring the `$CFLAGS` defined in the environment; they could use this instead.

```
$ unset CFLAGS
$ echo ${CFLAGS:--O2}
-O2
$ CFLAGS="-O3" echo ${CFLAGS:--O2}
-O2
$ export CFLAGS="-O3 -march=native"; echo ${CFLAGS:--O2}
-O3 -march=native

```

Maybe it's worth reaching out to the MPICH developers and ask why they don't?

**Comment by damianam** [comment]
I think it makes sense to have an easyblock to handle MPICH2 and all its derivatives compiled from source (which excludes Intel MPI, among others), and simply account for the differences between them. Alternatively, having an easyblock to handle MPICH2, and make the MVAPICH2 easyblock inhering from the new easyblock also looks appropriate.

The MVAPICH2 developers acknowledge the issue in their distributed RPMs and said it will be corrected in future releases.

The MPICH2 developers have been working like that for ages, so I am not sure if they'll be willing to change it. Even if they change it, it will take some time until it gets done, released, and cascades down to others like MVAPICH2 and PSMPI. I'd say we have to act at the EB level regardless of what they do.

Related to your last point: IMO the user should be responsible for how he/she compiles his/her code. In other words, if he/she defines `$CFLAGS`, but doesn't add it to the compilation line, maybe there is a reason for it. Maybe there is a script with multiple `mpicc` commands, of which just some of them need to have `$CFLAGS`. Adding it transparently seems like a mistake to me.

TL;DR:
* I support `MPICH2Derivatives` easyblocks, with `MVAPICH2` inhering from it.
* I oppose transparently adding `$CFLAGS` inside `mpixxx` in any of its forms.
* I'm neutral to requesting MPICH2 developers to do something more intuitive with their building process (i.e.: stop transparently embedding `$CFLAGS` at compile time in `mpixxx` wrappers)
* MVAPICH2 RPMs will not include undesired flags transparently in future releases.

**Comment by boegel** [comment]
@damianam: let's discuss this next week, @ocaisa mentioned that a single generic easyblock to cover all MPICH2-based MPI libraries may not work out well due to too many subtle differences between the different MPI libraries

**Comment by damianam** [comment]
@boegel the result of the discussion was that basically all derivatives compiled from source are still `./configure && make` packages. Pulling the configure options into the "base" MPICH easyblock is not a smart idea because of the differences. For instance ParaStationMPI has a multistaged configure step, where the options passed to initial configure are completely different than for MPICH or MVAPICH. However, the redefinition of the variables is the same for all of them. That, and the common files for the sanity checks is what has been pulled in the MPICH easyblock

**Comment by boegel** [comment]
This can be closed thanks to https://github.com/hpcugent/easybuild-easyblocks/pull/844?

**Comment by damianam** [comment]
I think it can :-)
